### DISCUSSION TITLE
Let User field to be required to be unique

### DISCUSSION THREAD

**Discussion by fubuloubu** [fief-dev/fief#384] [Ideas]
## Describe the bug

If you want to add a field that should be globally unique (within a tenent) such as a username, it is currently not possible to specify that

## To Reproduce

1. Go to 'User fields'
2. Click on 'Create User Field'
3. No way to add unique constraint on field

## Expected behavior

A checkbox in this form that adds a `UNIQUE` constraint on the field should be available, at least on some field types such as string, phone number, maybe address, etc.

## Configuration

* Cloud or self-hosted: self-hosted (Railway)
* If self-hosted, Fief version: 0.28.6

## Additional context

I want to protect the privacy of my user's emails by only allowing account search via username

**Comment by fief-bailiff** [comment]
Hail, @fubuloubu 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to [star our repository](https://github.com/fief-dev/fief) ⭐️

<p align="center">
<img width="656" alt="star-fief" src="https://github.com/fief-dev/fief/assets/1144727/b2db7f59-102f-47e5-9751-67fa10372c52">
</p>

## Want to support us?

Subscribe to one of our paid plan to help us continue our work and receive exclusive information and benefits! Starts at $5/month 🪙

<p align="center">
    <a href="https://polar.sh/fief-dev">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/embed/subscribe.svg?org=fief-dev&darkmode=1">
      <img alt="Subscribe" src="https://polar.sh/embed/subscribe.svg?org=fief-dev">
    </picture>
    </a>
</p>

Farewell!

**Comment by frankie567** [comment]
Sounds fair! Adding to the backlog :) https://github.com/orgs/fief-dev/projects/1/views/4?pane=issue&itemId=65301824

**Reply by fubuloubu** [reply]
What would it take to implement this? Would love to have this feature, one of my main blockers in using Fief in production right now

**Reply by frankie567** [reply]
I'm afraid we can't just add a unique constraint at DB level since we have a dynamic `user_field_values` table which represent the value for a given field and given user.

This would require to add some dedicated logic where we set the value, not that trivial.

**Reply by fubuloubu** [reply]
If it's difficult to enter into the DB still would be useful if the API could reject duplicates as a starting point
